File: dawn_cpp/ss.cpp
#include <algorithm>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

std::vector<std::string> get_ngrams(int n, const std::string& s) {
    std::vector<std::string> ngrams;
    for (int i = 0; i <= (int)s.length() - n; ++i) {
        ngrams.push_back(s.substr(i, n));
    }
    return ngrams;
}

std::unordered_map<std::string, int> count_3gram_frequencies(const std::vector<std::string>& sample_keys) {
    std::unordered_map<std::string, int> frequencies;
    for (const auto& key : sample_keys) {
        for (const auto& ngram : get_ngrams(3, key)) {
            ++frequencies[ngram];
        }
    }
    return frequencies;
}

std::vector<std::pair<std::string, std::optional<std::string>>> select_3gram_intervals(const std::vector<std::string>& sample_keys, int dict_size) {
    auto frequencies = count_3gram_frequencies(sample_keys);
    std::vector<std::pair<std::string, int>> sorted_frequencies(frequencies.begin(), frequencies.end());
    std::sort(sorted_frequencies.begin(), sorted_frequencies.end(), [](const auto& a, const auto& b) { return a.second > b.second; });

    std::vector<std::string> selected_ngrams;
    selected_ngrams.reserve(dict_size / 2);
    for (int i = 0; i < dict_size / 2; ++i) {
        selected_ngrams.push_back(sorted_frequencies[i].first);
    }
    std::sort(selected_ngrams.begin(), selected_ngrams.end());

    std::vector<std::pair<std::string, std::optional<std::string>>> intervals;
    std::string last_ngram;
    for (const auto& ngram : selected_ngrams) {
        std::string start = last_ngram;
        if (!start.empty()) {
            start.back() = start.back() + 1;
        }
        intervals.emplace_back(start, ngram);
        last_ngram = ngram;
    }
    return intervals;
}

File: dawn_rust/src/ss.rs
use std::collections::HashMap;

fn get_ngrams(n: usize, s: &str) -> Vec<&str> {
    let mut ngrams = Vec::new();
    for i in 0..s.len() - n + 1 {
        ngrams.push(&s[i..i + n]);
    }
    ngrams
}

fn count_3gram_frequencies(sample_keys: &[&str]) -> HashMap<&str, usize> {
    let mut frequencies = HashMap::new();
    for key in sample_keys {
        for ngram in get_ngrams(3, key) {
            *frequencies.entry(ngram).or_insert(0) += 1;
        }
    }
    frequencies
}

fn select_3gram_intervals(sample_keys: &[&str], dict_size: usize) -> Vec<(&str, Option<&str>)> {
    let mut frequencies = count_3gram_frequencies(sample_keys);
    let mut sorted_frequencies: Vec<_> = frequencies.iter().collect();
    sorted_frequencies.sort_by(|a, b| b.1.cmp(a.1));

    let selected_ngrams: Vec<_> = sorted_frequencies
        .iter()
        .take(dict_size / 2)
        .map(|(&ngram, _)| ngram)
        .collect();

    let mut intervals = Vec::new();
    let mut last_ngram = "";
    for ngram in selected_ngrams.iter().sorted() {
        let start = {
            let mut chars: Vec<_> = last_ngram.chars().collect();
            if !chars.is_empty() {
                chars.last_mut().unwrap().next_code();
            }
            chars.iter().collect()
        };
        intervals.push((start, Some(ngram)));
        last_ngram = ngram;
    }
    intervals
}